### ISSUE TITLE
[TF2 REQUEST] Custom Contracts

### ISSUE THREAD

**Issue by SirBlockles** [ValveSoftware/Source-1-Games#2720]
Apologies for making three long-winded posts, this is the last one (for now), I promise.

With the new ConTracker UI, Contracts have been made easy to follow and give something for more experienced players to do during the new updates. While there are server plugins that create "custom contracts," they don't tap into the native UI, and, at least from any public one i've seen, are very limited in what can be made an objective.

My proposal is to create a built-in custom Contract system, with two customization areas: client and server. Before I explain, let me share a mockup of the root ConTracker panel under my idea.
![contracker root](

For both client and server contracts, inside the directory would probably be preferable to have a scrollable list of subdirectories, to ensure that there aren't more contracts than can take up the screen. Inside these folders, though, there could be metadata for each contract node, giving its position, icon, arrow pointing to another node, etc., as well as general folder metadata such as the background image (similar to the unique background images for each folder of the stock ConTracker).

Under the client folder, any contracts added by the user themself will show up there. Folders inside this folder could possibly mirror the folder structure in whatever folder they end up in (`tf/contracts`?). User contracts would be able to be fulfillable in any server, and would offer no reward or acknowledgement upon completion. The only notification at all would possibly be the ConTracker action item making the progress tick noises off the player, which couldn't get any more annoying and spammy than the Noisemakers anyway. Client-side contracts would also have their progress stored locally in the player's data folder, but since there's zero reward for completing a client-side contract anyway, there's no real consequence to tampering with contract progress (in fact, having a command to manually increase contract progress for debugging purposes would probably serve well). The idea with client contracts is that they can be packaged up into VPKs or just put on a site like GameBanana, and "challenge packs" can be created and shared with others, to simply drop in your custom folder.

For server-side contracts, the contract map and progress would be stored on the server, which means that a client couldn't directly tamper with contract progress. Having contract progress readable by the server allows it to be manipulated by it too. This means that if there are contract triggers that aren't covered by the default contract system, an external server plugin like SourceMod can create and manage custom triggers. Monitoring contract progress can also lead to server-side rewards, such as someone completing all server-side contracts earning a free "VIP" rank, or other similar incentives, with the only investment required being time. Server-side contracts can also be good for "syndicated" servers (skial, no heroes, etc) by having a system to synchronize contract progress between all servers of a network, creating a sense of consistency. There would be a bit more flexibility on server-side contracts for recognition as well, for example allowing the "player has completed a bonus objective on their <contract> contract!" message to be broadcast, since it can't be spammed like a client-side contract could possibly be set up to be.

I considered writing about making contracts "party shared," where the party lead can mirror their contract onto their party members, but it would be simpler to just have all party members download the same client contracts, or do the same server contracts at the same time. However, the co-op mechanic used in the Merc Park contracts could make a return, though I don't know how the system is handled internally, and if it would take a lot of effort to listen to all of your party members' events and apply them to your own contracts. Maybe co-op contracts would be a server contract only thing, or would require both party members to have the same contract activated.

Client contracts would have no restriction on replay or "unlocking" via the star system, and progress on any contract could be reset at any time to do it over again. Server contracts may make this optional through CVAR, and may also be able to implement "unlocking" contracts, though that might get convoluted, especially if the same star system would be used (especially to make sure players don't get "trapped" not having enough stars to unlock any remaining contracts).

Internally, setting up contract nodes would be similar to a JSON or VDF file, with fields for the contract name, description, background banner, and objectives. Objectives would probably have a custom string for its objective display, while the criteria for the objective could be a combination of conditional triggers and events. Triggers would include things like kills, assists, healing, capping/pushing objectives, as well as class-specific things like ubers, building/demoing buildings, etc. Conditionals would be things like: playerclass, health, ammo, status effects (conds), map time, objective status (current points capped, KOTH point ownership, flag carrier status, etc). Conditionals would be able to be applied to the player, the target, and the assistor if, for example, the event was a kill or death.

Naturally, these contracts would use the native ConTracker UI, as well as the native contract progress UI. Server (and client?) contracts would also possibly be able to set a "reward" section to a contract node, with just a VTF defined, and would look similar to the stock reward section. This can be used in conjunction with server-side mods to show server-side rewards as mentioned earlier.

In conclusion, I think adding custom contracts to the game will allow people to come up with crazy challenges and breathe more "things to do" into the game with just, at its core, a simple progress bar that makes noise when you do things. Hell, you could market custom contracts, along with balance changes and maybe another cosmetic/warpaint case, as a "major update" and I would be happy. Thanks for reading.

**Comment by LoneLantern** [comment]
This is a great idea. What if we had contract packs to also be workshop items? Valve employees could pick 1-2 packs alongside a update-specific inhouse pack (which could add more options & events to the contracting system, since its made directly by the employee). Since you mentioned exporting to vpk(s) for sites like gamebanana, the workshop seems like the next logical step

**Comment by SirBlockles** [comment]
To be honest, if Valve even released more official contract passes like they did Merc Park (and i figure they kept the merc park contracts around because of how profitable it is), they could probably simultaneously make bank as well as give stuff for players to do.

Even if with every major update they just added more contracts to the merc park pack and charged only like $1 or $2 for people who already had the existing merc park contracts just to unlock the new ones, it would probably still give them quite a bit of profit without making the players feel cheated of money.

As for the workshop idea, i hadn't even thought of that! It could even work similarly to how Garry's Mod or SFM handles workshop content, where you just subscribe and it shows up in your list of contracts to do. That would make it even easier to install custom contracts for people who already don't quite know how to install customs in general, and with how the workshop works in gmod/SFM, it could even allow for version control with automatic updating. It's actually a really great idea, especially for the end user.
